### PR TITLE
[Windows] Add dimension fields for service datastream to support TSDB.

### PR DESCRIPTION
**Pull request by ritalwar**
- Enhancement

## What does this PR do?

This PR adds dimension fields for service datastream to support TSDB enablement.

## Checklist

- [ ] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [ ] I have verified that all data streams collect metrics or logs.
- [ ] I have added an entry to my package's `changelog.yml` file.
- [ ] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Related issues

- Relates #6993

### CODE REVIEW

**Comment by agithomas** [comment]
@ritalwar , there exist only one field - `uptime.ms`  having a metric value. Is there merit in doing TSDB enablement for this datastream? 

cc @lalit-satapathy

**Comment by lalit-satapathy** [comment]
> @ritalwar , there exist only one field - `uptime.ms` having a metric value. Is there merit in doing TSDB enablement for this datastream?
> 
> cc @lalit-satapathy

Single field metric document is fine. What metric type is _uptime.ms_? @ritalwar Can you share a sample service data stream document below?

**Comment by ritalwar** [comment]
> > @ritalwar , there exist only one field - `uptime.ms` having a metric value. Is there merit in doing TSDB enablement for this datastream?
> > cc @lalit-satapathy
> 
> Single field metric document is fine. What metric type is _uptime.ms_? @ritalwar Can you share a sample service data stream document below?

uptime is the time duration since the service was last started, and uptime.ms is the service's uptime specified in milliseconds.

Here is the sample doc:
```
{
  "_index": ".ds-metrics-windows.service-default-2023.07.28-000001",
  "_id": "459ztkAeenvETSm7AAABiZwAkRs",
  "_version": 1,
  "_score": 0,
  "_source": {
    "@timestamp": "2023-07-28T10:17:10.171Z",
    "agent": {
      "ephemeral_id": "609668fe-f058-47d4-ab34-e7a997423748",
      "id": "fc86843b-1617-4040-9201-67df23bc5222",
      "name": "obs-service-int",
      "type": "metricbeat",
      "version": "8.8.0"
    },
    "cloud": {
      "account": {
        "id": "0e073ec1-c22f-4488-adde-da35ed609ccd"
      },
      "instance": {
        "id": "ce7433e0-ce93-4944-b337-3492832fef60",
        "name": "obs-service-integeration-2019"
      },
      "machine": {
        "type": "Standard_D32ds_v5"
      },
      "provider": "azure",
      "region": "eastus",
      "service": {
        "name": "Virtual Machines"
      }
    },
    "data_stream": {
      "dataset": "windows.service",
      "namespace": "default",
      "type": "metrics"
    },
    "ecs": {
      "version": "8.0.0"
    },
    "elastic_agent": {
      "id": "fc86843b-1617-4040-9201-67df23bc5222",
      "snapshot": false,
      "version": "8.8.0"
    },
    "event": {
      "agent_id_status": "verified",
      "dataset": "windows.service",
      "duration": 62210800,
      "ingested": "2023-07-28T10:17:11Z",
      "module": "windows"
    },
    "host": {
      "architecture": "x86_64",
      "hostname": "obs-service-int",
      "id": "496839eb-5050-4edf-8b05-8df87de74876",
      "ip": [
        "10.7.0.5",
        "fe80::4b15:4076:808b:b97d"
      ],
      "mac": "00-0D-3A-53-23-DF",
      "name": "obs-service-int",
      "os": {
        "build": "17763.4645",
        "family": "windows",
        "kernel": "10.0.17763.4644 (WinBuild.160101.0800)",
        "name": "Windows Server 2019 Datacenter",
        "platform": "windows",
        "type": "windows",
        "version": "10.0"
      }
    },
    "metricset": {
      "name": "service",
      "period": 60000
    },
    "service": {
      "type": "windows"
    },
    "windows": {
      "service": {
        "display_name": "Distributed Transaction Coordinator",
        "id": "-VO2Lm89v4",
        "name": "MSDTC",
        "path_name": "C:\\Windows\\System32\\msdtc.exe",
        "pid": 11592,
        "start_name": "NT AUTHORITY\\NetworkService",
        "start_type": "Automatic (Delayed)",
        "state": "Running",
        "uptime": {
          "ms": 767641828
        }
      }
    }
  },
  "fields": {
    "elastic_agent.version": [
      "8.8.0"
    ],
    "host.os.name.text": [
      "Windows Server 2019 Datacenter"
    ],
    "host.hostname": [
      "obs-service-int"
    ],
    "host.mac": [
      "00-0D-3A-53-23-DF"
    ],
    "service.type": [
      "windows"
    ],
    "host.os.version": [
      "10.0"
    ],
    "host.os.name": [
      "Windows Server 2019 Datacenter"
    ],
    "windows.service.start_name": [
      "NT AUTHORITY\\NetworkService"
    ],
    "agent.name": [
      "obs-service-int"
    ],
    "host.name": [
      "obs-service-int"
    ],
    "event.agent_id_status": [
      "verified"
    ],
    "cloud.region": [
      "eastus"
    ],
    "host.os.type": [
      "windows"
    ],
    "windows.service.display_name": [
      "Distributed Transaction Coordinator"
    ],
    "data_stream.type": [
      "metrics"
    ],
    "host.architecture": [
      "x86_64"
    ],
    "cloud.machine.type": [
      "Standard_D32ds_v5"
    ],
    "cloud.provider": [
      "azure"
    ],
    "cloud.service.name": [
      "Virtual Machines"
    ],
    "agent.id": [
      "fc86843b-1617-4040-9201-67df23bc5222"
    ],
    "ecs.version": [
      "8.0.0"
    ],
    "windows.service.id": [
      "-VO2Lm89v4"
    ],
    "agent.version": [
      "8.8.0"
    ],
    "host.os.family": [
      "windows"
    ],
    "host.os.build": [
      "17763.4645"
    ],
    "host.ip": [
      "10.7.0.5",
      "fe80::4b15:4076:808b:b97d"
    ],
    "cloud.instance.id": [
      "ce7433e0-ce93-4944-b337-3492832fef60"
    ],
    "agent.type": [
      "metricbeat"
    ],
    "event.module": [
      "windows"
    ],
    "host.os.kernel": [
      "10.0.17763.4644 (WinBuild.160101.0800)"
    ],
    "windows.service.path_name": [
      "C:\\Windows\\System32\\msdtc.exe"
    ],
    "elastic_agent.snapshot": [
      false
    ],
    "windows.service.uptime.ms": [
      767641828
    ],
    "windows.service.state": [
      "Running"
    ],
    "host.id": [
      "496839eb-5050-4edf-8b05-8df87de74876"
    ],
    "windows.service.start_type": [
      "Automatic (Delayed)"
    ],
    "windows.service.pid": [
      11592
    ],
    "elastic_agent.id": [
      "fc86843b-1617-4040-9201-67df23bc5222"
    ],
    "data_stream.namespace": [
      "default"
    ],
    "metricset.period": [
      60000
    ],
    "metricset.name": [
      "service"
    ],
    "event.duration": [
      62210800
    ],
    "event.ingested": [
      "2023-07-28T10:17:11.000Z"
    ],
    "@timestamp": [
      "2023-07-28T10:17:10.171Z"
    ],
    "host.os.platform": [
      "windows"
    ],
    "cloud.account.id": [
      "0e073ec1-c22f-4488-adde-da35ed609ccd"
    ],
    "data_stream.dataset": [
      "windows.service"
    ],
    "agent.ephemeral_id": [
      "609668fe-f058-47d4-ab34-e7a997423748"
    ],
    "windows.service.name": [
      "MSDTC"
    ],
    "event.dataset": [
      "windows.service"
    ],
    "cloud.instance.name": [
      "obs-service-integeration-2019"
    ]
  }
}
```

File: packages/windows/data_stream/service/fields/fields.yml [hunk 1]
```diff
@@ -35,10 +36,12 @@
         For `Stopped` services this is the error code that service reports when starting to stopping. This will be the generic Windows service error code unless the service provides a service-specific error code.
     - name: pid
       type: long
+      #Reason to add as dimension field: Uniquely identify the running process, is important when service restarts and continues with the same service id.
```
Comment (agithomas) [inline packages/windows/data_stream/service/fields/fields.yml:39]: I read that these are the states of the services.`Continuing`, `Pausing`, `Paused`, `Running`, `Starting`, `Stopping`, and `Stopped`.

Does a stopped/paused service have a `pid` associated? If no, please do consider adding the `state` field additionally as a dimension.

Comment (ritalwar) [reply packages/windows/data_stream/service/fields/fields.yml:39]: > I read that these are the states of the services.`Continuing`, `Pausing`, `Paused`, `Running`, `Starting`, `Stopping`, and `Stopped`.
> 
> Does a stopped/paused service have a `pid` associated? If no, please do consider adding the `state` field additionally as a dimension.

pid is associated only with running processes.
Although in current testing scenarios, I have observed multiple services with different states and there is no change in the event count before and after enabling TSDB, do you still think having state as dimension is required?

Comment (agithomas) [reply packages/windows/data_stream/service/fields/fields.yml:39]: Among the states mentioned above, how many such states don't have the Pid? If it is more than one, you need to include the State as a dimension. Can you verify this please?

Comment (agithomas) [reply packages/windows/data_stream/service/fields/fields.yml:39]: Kindly also check, if the same service can be instantiated from multiple service account ? If yes, does the service share the same id?

Comment (ritalwar) [reply packages/windows/data_stream/service/fields/fields.yml:39]: > Kindly also check, if the same service can be instantiated from multiple service account ? If yes, does the service share the same id?

Yes, the same service can be instantiated from multiple service accounts, but each instantiation will have a different service id.

Comment (ritalwar) [reply packages/windows/data_stream/service/fields/fields.yml:39]: > Among the states mentioned above, how many such states don't have the Pid? If it is more than one, you need to include the State as a dimension, provided multiple instances of the same service can be instantiated at a given time. Can you verify this, please?

Yes, multiple instances of the same service can be instantiated at a given time. I will make state as dimension too.

**Review by agithomas — APPROVED** [review]
LGTM!

**Comment by lalit-satapathy** [comment]
> > 
> > Single field metric document is fine. What metric type is _uptime.ms_? @ritalwar Can you share a sample service data stream document below?


What metric type for uptime.ms?

**Comment by harnish-elastic** [comment]
@ritalwar, I have performed the TSDB-enablement-test-kit with this PR (including the metric_type field changes). The package is working fine. LGTM! 

image

**Review by harnish-elastic — APPROVED** [review]

**Pull request merged**